client/go/outline/config/parse.go

```diff
@@ -0,0 +1,139 @@
+func mapToAny(in map[string]any, out any) error {
+	newMap := make(map[string]any)
+	for k, v := range in {
+		if len(k) > 0 && k[0] == '$' {
+			// Skip $ keys
+			continue
+		}
+		newMap[k] = v
+	}
+	yamlText, err := yaml.Marshal(newMap)
+	if err != nil {
+		return fmt.Errorf("error marshaling to YAML: %w", err)
+	}
+	decoder := yaml.NewDecoder(bytes.NewReader(yamlText))
+	decoder.KnownFields(true)
+	if err := decoder.Decode(out); err != nil {
+		return fmt.Errorf("error decoding YAML: %w", err)
+	}
+	return nil
+}
+
+// TypeParser creates objects of the given type T from an input config.
+// You can register type-specific sub-parsers that get called when marked in the config.
+// The default value is not valid. Use [NewTypeParser] instead.
+type TypeParser[T any] struct {
+	fallbackHandler ParseFunc[T]
+	subparsers      map[string]func(context.Context, map[string]any) (T, error)
+}
+
+var _ ParseFunc[any] = (*TypeParser[any])(nil).Parse
+
+// NewTypeParser creates a [TypeParser] that calls the fallbackHandler if there's no parser specified in the config.
+func NewTypeParser[T any](fallbackHandler func(context.Context, ConfigNode) (T, error)) *TypeParser[T] {
+	return &TypeParser[T]{
+		fallbackHandler: fallbackHandler,
+		subparsers:      make(map[string]func(context.Context, map[string]any) (T, error)),
+	}
+}
+
+// Parse implements [ParseFunc] for the type T.
+func (p *TypeParser[T]) Parse(ctx context.Context, config ConfigNode) (T, error) {
+	var zero T
+
+	// Iterate while the input is a function call.
+	for {
+		inMap, ok := config.(map[string]any)
+		if !ok {
+			break
+		}
+		parserNameAny, ok := inMap[ConfigParserKey]
+		if !ok {
+			break
+		}
+		parserName, ok := parserNameAny.(string)
+		if !ok {
+			return zero, fmt.Errorf("parser name must be a string, found \"%T\"", parserNameAny)
+		}
+		parser, ok := p.subparsers[parserName]
+		if !ok {
+			return zero, fmt.Errorf("provider \"%v\" for type %T is not registered", parserName, zero)
+		}
+
+		// $parser is embedded in the value: {$parser: ..., ...}.
+		// Need to copy value and remove the type directive.
+		inputCopy := make(map[string]any, len(inMap))
+		for k, v := range inMap {
+			if k == ConfigParserKey {
+				continue
+			}
+			inputCopy[k] = v
+		}
+
+		var err error
+		config, err = parser(ctx, inputCopy)
+		if err != nil {
+			return zero, fmt.Errorf("parser \"%v\" failed: %w", parserName, err)
+		}
+	}
+
+	typed, ok := config.(T)
+	if ok {
+		return typed, nil
+	}
+
+	// Input is an intermediate type. We need a fallback handler.
+	return p.fallbackHandler(ctx, config)
+}
+
+// RegisterSubParser registers the given subparser function with the given name for the type T.
+// Note that a subparser always take a map[string]any, not ConfigNode, since we must have a map[string]any in
+// order to set the value for the ConfigParserKey.
+func (p *TypeParser[T]) RegisterSubParser(name string, function func(context.Context, map[string]any) (T, error)) {
```

Comment:
This function type is used in several places, shall we define a dedicated type for it?

Reply:
Defining the type won't help in anyway, since we are actually calling this method and have to write down the entire function declaration anyway for the anonymous function.